budget/app.py

```diff
@@ -106,8 +137,93 @@ def compare_budgets_and_users(account_id, users):
 
 def create_budget(synapse_id, team):
   '''Creates an AWS budget for a synapse user id'''
-  # TODO implement
-  log.info('Placeholder -- this will make a budget')
+  log.debug(f'Creating budget for synapse user {synapse_id}, member of team {team}')
+
+  team_budget_rules = TEAM_BUDGET_RULES.get('teams').get(team)
+  if not team_budget_rules:
+    raise ValueError(f'No budget rules available for team {team}')
+  budget_amount = team_budget_rules['amount']
+  budget_period = team_budget_rules['period']
+  budget = {
+    'BudgetName': f'{BUDGET_NAME_PREFIX}{synapse_id}',
+    'BudgetLimit': {
+      'Amount': budget_amount,
+      'Unit': 'USD'
+    },
+    'CostFilters': {
+      'TagKeyValue': [
+        (
+          'aws:servicecatalog:provisioningPrincipalArn$arn:aws:sts::'
+          f'{_get_account_id()}:assumed-role/ServiceCatalogEndusers/{synapse_id}'
+        )
+      ]
+    },
+    'CostTypes': {
+      'IncludeRefund': False,
+      'IncludeCredit': False
+    },
+    'TimeUnit': budget_period,
+    'BudgetType': 'COST'
+  }
+  budgets_client = get_client('budgets')
+  return budgets_client.create_budget(
+    AccountId=_get_account_id(),
+    Budget=budget
+    )
+
+
+def _create_budget_notification(synapse_id, threshold, admin_emails=None):
+  '''Creates a notification rule when the budget crosses a particular threshold.
+
+  May have multiple recipients. User notifications are sent through SNS
+  in order to send email to Synapse addresses, which cannot receive email
+  unless sent through the Synapse system, but emails to administrators can
+  be sent directly.
+  '''
+  budgets_client = get_client('budgets')
+  budget_name = f'{BUDGET_NAME_PREFIX}{synapse_id}'
+  subscribers = []
+  if admin_emails:
+    subscribers = [{
+      'SubscriptionType': 'EMAIL',
+      'Address': address
+    } for address in admin_emails]
+  subscribers.append({
+    'SubscriptionType': 'SNS',
+    'Address': _get_notification_topic_arn()
+    })
+
+  return budgets_client.create_notification(
+      AccountId=_get_account_id(),
+      BudgetName=budget_name,
+      Notification={
+          'NotificationType': 'ACTUAL',
+          'ComparisonOperator': 'GREATER_THAN',
+          'Threshold': threshold,
+          'ThresholdType': 'PERCENTAGE',
+          'NotificationState': 'ALARM'
+      },
+      Subscribers=subscribers
+  )
+
+
+def create_budget_notifications(synapse_id, team):
+  '''Creates notifications for a particular budget'''
+  budget_name = f'{BUDGET_NAME_PREFIX}{synapse_id}'
+  thresholds = TEAM_BUDGET_RULES['thresholds']
+
+  for threshold in thresholds['notify_user_only']:
+    _create_budget_notification(
```

Comment:
does this do anything right now?  seems like the `_create_budget_notification` only creates notification for admins right now?

Reply:
Incorrect. Lines 186-190 set up notifications for admins. Lines 191-194 set up notifications for users. I will add comments to make this more clear.

Reply:
OK. that is difficult to see without @brucehoff's SNS notification piece

Reply:
Again, I'm not sure where the confusion is. Bruce's lambda receives SNS messages -- the AWS Budgets service is what actually makes the notifications and sends to the SNS topic. Bruce's lambda exists because the Synapse system will not allow emails to be sent to synapse.org addresses from outside the Synapse system -- he has to send it through the Synapse API.

Reply:
yep, i get it now.  It was difficult to see how sending to an SNS topic would send notifications to individual synapse users.   For some reason the comments in the method header didn't click for me either.  might help if that comment was moved right above `subscribers.append(..` call.